README.org
#+TITLE: pueue

[[https://melpa.org/#/pueue][file:https://melpa.org/packages/pueue-badge.svg]] [[https://stable.melpa.org/#/pueue][file:https://stable.melpa.org/packages/pueue-badge.svg]]

This package provides an intuitive GUI for [[https://github.com/Nukesor/pueue][pueue]] task manager.

[[scrot.png]]

* Installation

** Nix

This repo is a nix flake.  You can just add ~overlays.default~ output to your
nixpkgs overlays.

#+begin_src nix
{
  inputs.pueue.url = "github:xFA25E/pueue";
  outputs = { self, pueue, nixpkgs }: let
    pkgs = import nixpkgs {
      system = "your-system";
      overlays = [ pueue.overlays.default ];
    };
  in {
    # YOUR OUTPUTS
  };
}
#+end_src

** Package manager

If you've installed it with your package manager, you're done.  ~pueue~ command
is autoloaded, so you can call it right away.

** Manual

It depends on the following packages:

+ with-editor

Then put /pueue.el/ file in your ~load-path~, and put the following in your init
file:

#+begin_src elisp
(require 'pueue)
#+end_src

* Usage

Run ~pueue~ command.  You can press ~?~ key or call ~describe-mode~ to discover
available keybindings.  Dired-like marking also works.

** Tips

+ You can customize settings in the *pueue* group.

+ Required version of pueue is 2.0.0.  Although the visualization of tasks and
  detailed view will work with 1.0.0, some commands (like group) have changed
  their API since 2.0.0.

* COMMENT Development

* License

GPLv3


CHANGELOG.org
* Changelog

All notable changes to this project will be documented in this file.

The format is based on [[https://keepachangelog.com/en/1.0.0/][Keep a Changelog]],
and this project adheres to [[https://semver.org/spec/v2.0.0.html][Semantic Versioning]].

** [2.1.1] - Unreleased

*** Added

*** Changed

*** Deprecated

*** Removed

*** Fixed
- ~pueue-group-remove~ command.

*** Security

** [2.1.0] - 2022-11-05

*** Added

- Customizable ~pueue-command~

*** Fixed

- After ~pueue-remove~ marked tasks are correctly unmarked

** [2.0.0] - 2022-09-08

*** Added

- Support for *pueue 2.0.0*.
- *pueue* edit commands are now launched with ~with-editor~ library.  User can
  press ~C-x #~ to finish editing.

*** Changed

- Renamed all *pueue* commands (~pueue-command-<name> => pueue-<name>~).
- Move everything to single file.
- ~pueue--call~ arguments are now flattened with ~flatten-tree~.

*** Removed

- ~pueue-info-mode~ is no longer implemented with *ewoc* library.
- Support for *pueue 1.0.0*.
- Support for *GNU/Emacs 27.1*.  *28.1* is required now.

** [1.0.4] - 2021-12-07

*** Added

- Group column.
- Group filters.
- Completion for some transient infixes.

*** Changed

- Make ~shell.nix~ more reproducible.
- Use ~easy-mmode-defmap~ instead of ~define-key~.
- Refactor ~pueue--refresh~.

*** Fixed

- Drawing of task status.

** [1.0.3] - 2021-11-08

*** Added

- ~pueue-command-help~ transient command that displays all available commands.

** [1.0.2] - 2021-11-04

*** Changed

- Refactor insertion and mark operations.

** [1.0.1] - 2021-11-03

*** Changed

- *pueue* ~log~ command is now implemented with ~async-shell-command~.

** [1.0.0] - 2021-11-03

*** Added

- Transient commands for every *pueue* subcommand.

*** Changed

- ~pueue-mode~ is now implemented with *tabulated-list* library.
- ~pueue-info-mode~ is now implemented with *ewoc* library.

*** Removed

- *bui* is no longer used.

** [0.1.0] - 2021-04-09

*** Added

- First working implementation with *bui*.


pueue-info.el
;;; pueue-info.el --- Detailed view of Pueue tasks   -*- lexical-binding: t; -*-

;; Copyright (C) 2021  Valeriy Litkovskyy

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; Detailed view for Pueue tasks.

;;; Code:

;;;; REQUIRES

(require 'ewoc)
(require 'iso8601)
(require 'map)

(defvar pueue--status)
(declare-function pueue--marked-ids "peueu")

;;;; CUSTOMZATION OPTIONS

;;;;; VARIABLES

(defcustom pueue-info-buffer-name
  "*Pueue Info*"
  "Default buffer name for Pueue Info mode."
  :type 'string
  :group 'pueue)

;;;;; FACES

(defface pueue-info-label
  '((t :inherit font-lock-constant-face))
   "Pueue Info mode face used for field labels."
   :group 'pueue)

(defface pueue-info-environment-variable
  '((t :inherit font-lock-variable-name-face))
   "Pueue Info mode face used for environment variables."
   :group 'pueue)

;;;; LOCAL VARIABLES

(defvar-local pueue-info--ewoc nil
  "Pueue Info mode ewoc data.")

(defvar-local pueue-info--tasks nil
  "Pueue tasks.
It is set during the `pueue-info' call.  It is a hash-table with
string ids as keys and hash-tables describing tasks as values.
Each individual task has fields and values.  Fields are strings
and values have the following types:

 \"id\"               -> number
 \"command\"          -> string
 \"label\"            -> string or nil
 \"path\"             -> string
 \"status\"           -> string or hash-table
 \"start\"            -> iso8601 time string
 \"end\"              -> iso8601 time string
 \"group\"            -> string
 \"enqueue_at\"       -> iso8601 time string or nil
 \"dependencies\"     -> vector of numbers
 \"prev_status\"      -> string
 \"original_command\" -> string
 \"envs\"             -> hash-table {string: string}")

(defvar-local pueue-info--history nil
  "List containing history of visited tasks.
History is recorded only when tasks are visited from the same
buffer.  Every time `pueue-info' is called from different buffer,
history is reset.

Every history item has the following form:

(((ID . DRAW-ENVS-P)...) . POSITION)

See `pueue-info--draw-envs' for more information on
DRAW-ENVS-P.")

;;;; UTILS

(defun pueue-info--footer-p (node)
  "Check whether NODE is ewoc footer."
  (eq (ewoc--footer pueue-info--ewoc) node))

(defun pueue-info--make-node-visible (node)
  "Make ewoc NODE visible as much as possible."
  (let* ((ewoc pueue-info--ewoc)
         (next-node (or (ewoc-next ewoc node) (ewoc--footer ewoc)))
         (next-node-location (ewoc-location next-node)))
    (unless (pos-visible-in-window-p next-node-location)
      (let ((node-line-number (line-number-at-pos (ewoc-location node)))
            (next-node-line-number (line-number-at-pos next-node-location)))
        (recenter (- (min (- next-node-line-number node-line-number) (window-height))))))))

;;;; DRAW FUNCTIONS

(defun pueue-info--draw-number (number &optional _)
  "Draw NUMBER value."
  (insert " " (number-to-string number)))

(defun pueue-info--draw-string (string &optional _)
  "Draw STRING value.
STRING can be a string or nil."
  (when string (insert " " string)))

(defun pueue-info--draw-time (time &optional _)
  "Draw iso8601 time string.
TIME can be a string or nil."
  (when time
    (seq-let (S M H d m y) (iso8601-parse time)
      (insert (format " %d-%02d-%02d %02d:%02d:%02d" y m d H M S)))))

(defun pueue-info--draw-result (result &optional _)
  "Draw pueue task RESULT.
RESULT can be a string or a hash-table."
  (pcase result
    ((pred stringp) (pueue-info--draw-string result))
    ((pred mapp)
     (pcase (seq-first (map-pairs result))
       (`("Failed" . ,code)
        (insert " Failed")
        (pueue-info--draw-number code))
       (`("FailedToSpawn" . ,reason)
        (insert " FailedToSpawn")
        (pueue-info--draw-string reason))))))

(defun pueue-info--draw-status (status &optional _)
  "Draw pueue task STATUS.
STATUS can be a string or a hash-table."
  (pcase status
    ((pred stringp) (pueue-info--draw-string status))
    ((pred mapp)
     (pcase (seq-first (map-pairs status))
       (`("Done" . ,result)
        (insert " Done")
        (pueue-info--draw-result result))
       (`("Stashed" . ,(map ("enqueue_at" enqueue-at)))
        (insert " Stashed")
        (pueue-info--draw-time enqueue-at))))))

(defun pueue-info--push-dependency-button (pos)
  "Push button on task dependency at position POS.
This function is used as action in dependency buttons.  It
records history to `pueue-info--history'."
  (let ((data (ewoc-collect pueue-info--ewoc #'identity))
        (task-id (button-get (button-at pos) 'task-id)))
    (push (cons data (point)) pueue-info--history)
    (ewoc-filter pueue-info--ewoc #'ignore)
    (ewoc-enter-last pueue-info--ewoc (cons task-id nil))))

(defun pueue-info--draw-dependencies (dependencies &optional _)
  "Draw DEPENDENCIES.
DEPENDENCIES is a vector of numbers.  Each number represents an
id of some pueue task.  DEPENDENCIES are drawn as clickable
buttons."
  (let ((action #'pueue-info--push-dependency-button))
    (seq-do (pcase-lambda ((and (app number-to-string id) task-id))
              (insert " ")
              (insert-text-button id 'action action 'task-id task-id))
            dependencies)))

(defun pueue-info--draw-envs (envs draw-envs-p)
  "Draw pueue task environment variables.
ENVS is a hash-table {string: string} and DRAW-ENVS-P is a
boolean.  It will draw environment variables only if DRAW-ENVS-P
is non-nil."
  (if (not draw-envs-p)
      (insert "...")
    (insert "\n")
    (seq-do (pcase-lambda (`(,var . ,value))
              (let ((var (propertize var 'face 'pueue-info-environment-variable)))
                (insert "  " var "=" value "\n")))
            (seq-sort-by #'car #'string< (map-pairs envs)))))

(defun pueue-info--draw-task (data)
  "Draw pueue TASK.
This is a main function used as ewoc pretty printer.  DATA has
the following form: (ID . DRAW-ENVS-P).  ID is a number and
DRAW-ENVS-P is a boolean.

See `pueue-info--draw-envs' for more information on DRAW-ENVS-P."
  (pcase-let* ((`(,(app number-to-string id) . ,draw-envs-p) data)
               ((map (id task)) pueue-info--tasks))
    (seq-do
     (pcase-lambda ((seq key label fn))
       (insert (propertize label 'face 'pueue-info-label) ":")
       (funcall fn (map-elt task key) draw-envs-p)
       (insert "\n"))
     [["id" "ID" pueue-info--draw-number]
      ["command" "Command" pueue-info--draw-string]
      ["label" "Label" pueue-info--draw-string]
      ["path" "Path" pueue-info--draw-string]
      ["status" "Status" pueue-info--draw-status]
      ["start" "Start" pueue-info--draw-time]
      ["end" "End" pueue-info--draw-time]
      ["group" "Group" pueue-info--draw-string]
      ["enqueue_at" "Enqueue at" pueue-info--draw-time]
      ["dependencies" "Dependencies" pueue-info--draw-dependencies]
      ["prev_status" "Previous status" pueue-info--draw-status]
      ["original_command" "Original command" pueue-info--draw-string]
      ["envs" "Environment variables" pueue-info--draw-envs]])))

;;;; BINDINGS

(easy-mmode-defmap pueue-info-mode-map
  '(("n" . pueue-info-next-task)
    ("p" . pueue-info-previous-task)
    ("e" . pueue-info-toggle-envs)
    ("l" . pueue-info-backward-history))
  "Keymap for `pueue-info-mode'.")

;;;; COMMANDS

(define-derived-mode pueue-info-mode special-mode "PueueInfo"
  "Mode used to draw details pueue tasks information."
  :group 'pueue
  (buffer-disable-undo)
  (with-silent-modifications (erase-buffer))
  (setq pueue-info--ewoc (ewoc-create #'pueue-info--draw-task)))

(defun pueue-info (tasks task-ids)
  "Main entry command to display details of pueue TASK-IDS.
TASK-IDS is a sequence of numbers.  Each number corresponds to
some pueue task id in TASKS.

When this command is called from Pueue mode, it gathers
automatically all TASKS in that buffer.

See `pueue-info--tasks' for more information on TASKS."
  (interactive
   (if (not (derived-mode-p 'pueue-mode))
       (user-error "Not in Pueue mode")
     (list (map-elt pueue--status "tasks") (pueue--marked-ids))))

  (with-current-buffer (get-buffer-create pueue-info-buffer-name)
    (pueue-info-mode)
    (setq pueue-info--tasks tasks)
    (seq-doseq (task-id task-ids)
      (ewoc-enter-last pueue-info--ewoc (cons task-id nil)))
    (pop-to-buffer (current-buffer))))

(defun pueue-info-next-task (n)
  "Go to next Nth task."
  (interactive "p")
  (let ((node (ewoc-goto-next pueue-info--ewoc n)))
    (unless (pueue-info--footer-p node)
      (pueue-info--make-node-visible node))))

(defun pueue-info-previous-task (n)
  "Go to previous Nth taks."
  (interactive "p")
  (ewoc-goto-prev pueue-info--ewoc n))

(defun pueue-info-toggle-envs ()
  "Toggle display of environment variables of a task at point."
  (interactive)
  (when-let ((node (ewoc-locate pueue-info--ewoc)))
    (let ((data (ewoc-data node)))
      (setcdr data (not (cdr data))))
    (ewoc-invalidate pueue-info--ewoc node)
    (pueue-info--make-node-visible node)))

(defun pueue-info-backward-history ()
  "Go back in history."
  (interactive)
  (when-let ((item (pop pueue-info--history)))
    (ewoc-filter pueue-info--ewoc #'ignore)
    (mapc (apply-partially #'ewoc-enter-last pueue-info--ewoc) (car item))
    (goto-char (cdr item))))

;;;; PROVIDE

(provide 'pueue-info)
;;; pueue-info.el ends here


pueue-command.el
;;; pueue-command.el --- Pueue interactive commands  -*- lexical-binding: t; -*-

;; Copyright (C) 2021  Valeriy Litkovskyy

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; Interactive pueue commands

;;; Code:

;;;; REQUIRES

(require 'map)
(require 'transient)

(defvar crm-completion-table)
(declare-function pueue--marked-ids "pueue")
(declare-function pueue--status "pueue")

;;;; VARIABLES

(defvar pueue-command--delay-help (string-trim "
2020-04-01T18:30:00   # RFC 3339 timestamp           | wednesday # The closest wednesday in the future
2020-4-1 18:2:30      # Optional leading zeros       | 4 months  # 4 months from today at 00:00:00
2020-4-1 5:30pm       # Informal am/pm time          | 1 week    # 1 week at the current time
2020-4-1 5pm          # Optional minutes and seconds | 1days     # 1 day from today at the current time
April 1 2020 18:30:00 # English months               | 1d 03:00  # The closest 3:00 after 1 day (24 hours)
1 Apr 8:30pm          # Implies current year         | 3h        # 3 hours from now
4/1                   # American form date           | 3600s     # 3600 seconds from now
wednesday 10:30pm     # The closest wednesday in the future at 22:30
") "Delay minibuffer help.")

;;;; COMMANDS

(defun pueue-command--call (command &optional args)
  "Call pueue COMMAND with ARGS.
Show success or error message."
  (with-temp-buffer
    (apply #'call-process "pueue" nil t nil command args)
    (message "%s" (string-trim-right (buffer-string)))))

(defun pueue-command--completing-read-group (prompt initial-input history)
  "Completing read pueue group.
PROMPT, INITIAL-INPUT and HISTORY are the same as in
`completing-read'."
  (let ((groups (map-keys (map-elt (pueue--status) "groups"))))
    (completing-read prompt groups nil t initial-input history)))

(defun pueue-command--completing-read-multiple-pueue-tasks (prompt initial-input history)
  "Completing read multiple pueue tasks.
PROMPT, INITIAL-INPUT and HISTORY are the same as in
`completing-read-multiple'."
  (let* ((tasks (map-elt (pueue--status) "tasks"))
         (completion-extra-properties
          (list :annotation-function
                (lambda (id)
                  (when-let ((task (map-elt crm-completion-table id)))
                    (concat "  -- " (map-elt task "command"))))))
         (ids (completing-read-multiple prompt tasks nil t initial-input history)))
    (string-join ids ",")))

(defun pueue-command--completing-read-delay (prompt initial-input history)
  "Completing read delay argument.
It adds minibuffer help.  PROMPT, INITIAL-INPUT and HISTORY are
the same as in `completing-read'."
  (let ((minibuffer-help-form pueue-command--delay-help))
    (read-from-minibuffer prompt initial-input nil nil history)))

;;;;; ADD

(defun pueue-command-add--transient (command &optional args)
  "Transient helper.
Run pueue add command.  COMMAND is a string and ARGS are command
line arguments."
  (interactive
   (list (read-shell-command "Command: ")
         (mapcan
          (lambda (arg)
            (if (string-match-p (rx bos "--after=") arg)
                (mapcar
                 (apply-partially #'concat "--after=")
                 (split-string (substring arg 8) ","))
              (list arg)))
          (transient-args 'pueue-command-add))))
  (pueue-command--call "add" (append args (list command)))
  (revert-buffer nil t))

(transient-define-prefix pueue-command-add ()
  "Run pueue add command."
  ["Flags"
   ("-e" "Escape any special shell characters" "--escape")
   ("-i" "Immediately start the task" "--immediate")
   ("-s" "Create the task in Stashed state" "--stashed")]
  ["Options"
   ("-a" "Start the task after other tasks" "--after=" pueue-command--completing-read-multiple-pueue-tasks)
   ("-d" "Prevents the task from being enqueued until <delay> elapses" "--delay=" pueue-command--completing-read-delay)
   ("-g" "Assign the task to a group" "--group=" pueue-command--completing-read-group)
   ("-l" "Add some information for yourself" "--label=")]
  ["Actions"
   ("a" "Add" pueue-command-add--transient)])

;;;;; CLEAN

(defun pueue-command-clean--transient (&optional args)
  "Transient helper.
Run pueue clean command.  ARGS are command line arguments."
  (interactive (list (transient-args 'pueue-command-clean)))
  (pueue-command--call "clean" args)
  (revert-buffer nil t))

(transient-define-prefix pueue-command-clean ()
  "Run pueue clean command."
  ["Flags"
   ("-s" "Only clean tasks that finished successfully" "--successful-only")]
  ["Actions"
   ("c" "Clean" pueue-command-clean--transient)])

;;;;; EDIT

(defun pueue-command-edit--transient (task-id &optional args)
  "Transient helper.
Run pueue edit command.  TASK-ID is a numeric id of a task.  ARGS
are command line arguments."
  (interactive
   (let ((marked-ids (pueue--marked-ids)))
     (if (or (not marked-ids) (/= 1 (seq-length marked-ids)))
         (user-error "No task id or more than one task id specified")
       (list (seq-first marked-ids) (transient-args 'pueue-command-edit)))))
  (pueue-command--call "edit" (append (list (number-to-string task-id)) args))
  (revert-buffer nil t))

(transient-define-prefix pueue-command-edit ()
  "Run pueue edit command."
  ["Flags"
   ("-p" "Edit the path of the task" "--path")]
  ["Actions"
   ("e" "Edit" pueue-command-edit--transient)])

;;;;; ENQUEUE

(defun pueue-command-enqueue--transient (task-ids &optional args)
  "Transient helper.
Run pueue enqueue command.  TASK-IDS are numeric task ids.  ARGS
are command line arguments."
  (interactive (list (pueue--marked-ids) (transient-args 'pueue-command-enqueue)))
  (let ((ids (seq-map #'number-to-string task-ids)))
    (pueue-command--call "enqueue" (append args ids))
    (revert-buffer nil t)))

(transient-define-prefix pueue-command-enqueue ()
  "Run pueue enqueue command."
  ["Options"
   ("-d" "Delay enqueuing these tasks until <delay> elapses" "--delay=" pueue-command--completing-read-delay)]
  ["Actions"
   ("Q" "Enqueue" pueue-command-enqueue--transient)])

;;;;; FOLLOW

(defun pueue-command-follow--transient (task-id &optional args)
  "Transient helper.
Run pueue follow command.  TASK-ID is a numeric id of a task.
ARGS are command line arguments."
  (interactive
   (let ((marked-ids (pueue--marked-ids)))
     (if (or (not marked-ids) (/= 1 (seq-length marked-ids)))
         (user-error "No task id or more than one task id specified")
       (list (seq-first marked-ids) (transient-args 'pueue-command-follow)))))

  (let* ((buffer-name "*Pueue Follow*")
         (id (number-to-string task-id))
         (command (append (list "pueue" "follow") args (list id))))
    (when-let ((buffer (get-buffer buffer-name)))
      (kill-buffer buffer))
    (async-shell-command (string-join command " ") buffer-name)))

(transient-define-prefix pueue-command-follow ()
  "Run pueue follow command."
  ["Flags"
   ("-e" "Show stderr instead of stdout" "--err")]
  ["Actions"
   ("f" "Follow" pueue-command-follow--transient)])

;;;;; GROUP

(defun pueue-command-group--transient (&optional args)
  "Transient helper.
Run pueue group command.  ARGS are command line arguments."
  (interactive (list (transient-args 'pueue-command-group)))
  (pueue-command--call "group" args)
  (revert-buffer nil t))

(transient-define-prefix pueue-command-group ()
  "Run pueue group command."
  ["Options"
   ("-a" "Add a group by name" "--add=")
   ("-r" "Remove a group by name" "--remove=" pueue-command--completing-read-group)]
  ["Actions"
   ("G" "Group" pueue-command-group--transient)])

;;;;; KILL

(defun pueue-command-kill--transient (task-ids &optional args)
  "Transient helper.
Run pueue kill command.  TASK-IDS are numeric ids of tasks.  ARGS
are command line arguments."
  (interactive (list (pueue--marked-ids) (transient-args 'pueue-command-kill)))
  (let ((ids (seq-map #'number-to-string task-ids)))
    (pueue-command--call "kill" (append args ids))
    (revert-buffer nil t)))

(transient-define-prefix pueue-command-kill ()
  "Run pueue kill command."
  ["Flags"
   ("-a" "Kill all running tasks across ALL groups. This also pauses all groups" "--all")
   ("-c" "Send the SIGTERM signal to all children as well" "--children")]
  ["Options"
   ("-g" "Kill all running tasks in a group. This also pauses the group" "--group="
    pueue-command--completing-read-group)]
  ["Actions"
   ("k" "Kill" pueue-command-kill--transient)])

;;;;; LOG

(defun pueue-command-log--transient (task-ids &optional args)
  "Transient helper.
Run pueue log command.  TASK-IDS are numeric ids of tasks.  ARGS
are command line arguments."
  (interactive (list (pueue--marked-ids) (transient-args 'pueue-command-log)))
  (let* ((ids (seq-map #'number-to-string task-ids))
         (command (append (list "pueue" "log") args ids))
         (buffer-name "*Pueue Log*"))
    (async-shell-command (string-join command " ") buffer-name)))

(transient-define-prefix pueue-command-log ()
  "Run pueue log command."
  ["Flags"
   ("-f" "Show the whole stdout and stderr output" "--full")]
  ["Options"
   ("-l" "Only print the last X lines of each task's output" "--lines=")]
  ["Actions"
   ("l" "Log" pueue-command-log--transient)])

;;;;; PARALLEL

(defun pueue-command-parallel--transient (parallel-tasks &optional args)
  "Transient helper.
Run pueue parallel command.  PARALLEL-TASKS is a number.  ARGS
are command line arguments."
  (interactive (list (read-number "Parallel tasks: ")
                     (transient-args 'pueue-command-parallel)))
  (let ((parallel-tasks (number-to-string parallel-tasks)))
    (pueue-command--call "parallel" (append args (list parallel-tasks)))
    (revert-buffer nil t)))

(transient-define-prefix pueue-command-parallel ()
  "Run pueue parallel command."
  ["Options"
   ("-g" "Set the amount for the specific group" "--group="
    pueue-command--completing-read-group)]
  ["Actions"
   ("L" "Parallel" pueue-command-parallel--transient)])

;;;;; PAUSE

(defun pueue-command-pause--transient (task-ids &optional args)
  "Transient helper.
Run pueue pause command.  TASK-IDS are numeric ids of tasks.
ARGS are command line arguments."
  (interactive (list (pueue--marked-ids) (transient-args 'pueue-command-pause)))
  (let ((ids (seq-map #'number-to-string task-ids)))
    (pueue-command--call "pause" (append args ids))
    (revert-buffer nil t)))

(transient-define-prefix pueue-command-pause ()
  "Run pueue pause command."
  ["Flags"
   ("-a" "Pause all groups!" "--all")
   ("-c" "Also pause direct child processes of a task's main proces." "--children")]
  ["Options"
   ("-g" "Pause a specific group" "--group="
    pueue-command--completing-read-group)]
  ["Actions"
   ("P" "Pause" pueue-command-pause--transient)])

;;;;; REMOVE

(defun pueue-command-remove (task-ids)
  "Run pueue remove command.
TASK-IDS are numeric ids of tasks."
  (interactive (list (pueue--marked-ids)))
  (let ((ids (seq-map #'number-to-string task-ids)))
    (pueue-command--call "remove" ids)
    (revert-buffer nil t)))

;;;;; RESET

(defun pueue-command-reset--transient (&optional args)
  "Transient helper.
Run pueue reset command.  ARGS are command line arguments."
  (interactive (list (transient-args 'pueue-command-reset)))
  (pueue-command--call "reset" args)
  (revert-buffer nil t))

(transient-define-prefix pueue-command-reset ()
  "Run pueue reset command."
  ["Flags"
   ("-c" "Send the SIGTERM signal to all children as well" "--children")
   ("-f" "Don't ask for any confirmation" "--force")]
  ["Actions"
   ("t" "Reset" pueue-command-reset--transient)])

;;;;; RESTART

(defun pueue-command-restart--transient (task-ids &optional args)
  "Run pueue restart command.
TASK-IDS are numeric ids of tasks.  ARGS are command line
arguments."
  (interactive (list (pueue--marked-ids) (transient-args 'pueue-command-restart)))
  (let ((ids (seq-map #'number-to-string task-ids)))
    (pueue-command--call "restart" (append args ids))
    (revert-buffer nil t)))

(transient-define-prefix pueue-command-restart ()
  "Run pueue restart command."
  ["Flags"
   ("-a" "Restart all failed tasks" "--all-failed")
   ("-e" "Edit the tasks' command before restarting" "--edit")
   ("-p" "Edit the tasks' path before restarting" "--edit-path")
   ("-i" "Restart the task by reusing the already existing tasks" "--in-place")
   ("-k" "Immediately start the tasks" "--start-immediately")
   ("-s" "Set the restarted task to a \"Stashed\" state" "--stashed")]
  ["Actions"
   ("r" "Restart" pueue-command-restart--transient)])

;;;;; SEND

(defun pueue-command-send (task-id input)
  "Run pueue send command.
TASK-ID is a numeric id of task.  INPUT is a string."
  (interactive
   (let ((marked-ids (pueue--marked-ids)))
     (if (or (not marked-ids) (/= 1 (seq-length marked-ids)))
         (user-error "No task id or more than one task id specified")
       (list (seq-first marked-ids) (read-string "Input: ")))))
  (let ((id (number-to-string task-id)))
    (pueue-command--call "send" (list id input))
    (revert-buffer nil t)))

;;;;; START

(defun pueue-command-start--transient (task-ids &optional args)
  "Transient helper.
Run pueue start command.  TASK-IDS are numeric ids of tasks.
ARGS are command line arguments."
  (interactive (list (pueue--marked-ids) (transient-args 'pueue-command-start)))
  (let ((ids (seq-map #'number-to-string task-ids)))
    (pueue-command--call "start" (append args ids))
    (revert-buffer nil t)))

(transient-define-prefix pueue-command-start ()
  "Run pueue start command."
  ["Flags"
   ("-a" "Resume all groups" "--all")
   ("-c" "Also resume direct child processes of your paused tasks" "--children")]
  ["Options"
   ("-g" "Resume a specific group and all paused tasks in it" "--group="
    pueue-command--completing-read-group)]
  ["Actions"
   ("s" "Start" pueue-command-start--transient)])

;;;;; STASH

(defun pueue-command-stash (task-ids)
  "Run pueue stash command.
TASK-IDS are numeric ids of tasks."
  (interactive (list (pueue--marked-ids)))
  (let ((ids (seq-map #'number-to-string task-ids)))
    (pueue-command--call "stash" ids)
    (revert-buffer nil t)))

;;;;; SWITCH

(defun pueue-command-switch (task-id-1 task-id-2)
  "Run pueue switch command.
TASK-ID-1 and TASK-ID-2 are numeric ids of tasks."
  (interactive
   (let ((marked-ids (pueue--marked-ids)))
     (if (or (not marked-ids) (/= 2 (seq-length marked-ids)))
         (user-error "No task id or not two task ids specified")
       marked-ids)))
  (let ((ids (mapcar #'number-to-string (list task-id-1 task-id-2))))
    (pueue-command--call "switch" ids))
  (revert-buffer nil t))

;;;;; HELP

(transient-define-prefix pueue-command-help ()
  "Show all pueue commmands."
  [["Task actions"
    ("e" "Edit" pueue-command-edit)
    ("Q" "Enqueue" pueue-command-enqueue)
    ("k" "Kill" pueue-command-kill)
    ("P" "Pause" pueue-command-pause)]
   [""
    ("R" "Remove" pueue-command-remove)
    ("r" "Restart" pueue-command-restart)
    ("s" "Start" pueue-command-start)
    ("H" "Stash" pueue-command-stash)]
   ["Task info"
    ("f" "Follow" pueue-command-follow)
    ("l" "Log" pueue-command-log)
    ("d" "Send" pueue-command-send)]
   ["Queue actions"
    ("a" "Add" pueue-command-add)
    ("c" "Clean" pueue-command-clean)
    ("w" "Switch" pueue-command-switch)]
   [""
    ("G" "Group" pueue-command-group)
    ("L" "Parallel" pueue-command-parallel)
    ("t" "Reset" pueue-command-reset)]])

;;;; PROVIDE

(provide 'pueue-command)
;;; pueue-command.el ends here
